poc_iot_injector/src/server.rs

```diff
@@ -99,36 +102,56 @@ impl Server {
 }
 
 async fn submit_txns(
-    txn_service: &mut TransactionService,
+    txn_service: Arc<Mutex<TransactionService>>,
     store: &mut FileStore,
-    keypair: &Keypair,
+    keypair: Arc<Keypair>,
     after_utc: DateTime<Utc>,
     before_utc: DateTime<Utc>,
 ) -> Result {
     let file_list = store
         .list_all(FileType::LoraValidPoc, after_utc, before_utc)
         .await?;
 
-    let mut stream = store.source(stream::iter(file_list).map(Ok).boxed());
-    let ts = before_utc.timestamp_millis();
+    let stream = store.source(stream::iter(file_list).map(Ok).boxed());
+    let before_ts = before_utc.timestamp_millis();
 
-    while let Some(Ok(msg)) = stream.next().await {
-        if let Ok(Some((txn, hash, hash_b64_url))) = handle_report_msg(msg, keypair, ts) {
-            if txn_service
-                .submit(
-                    BlockchainTxn {
-                        txn: Some(Txn::PocReceiptsV2(txn)),
-                    },
-                    &hash,
-                )
-                .await
-                .is_ok()
-            {
-                tracing::debug!("txn submitted successfully, hash: {:?}", hash_b64_url);
-            } else {
-                tracing::warn!("txn submission failed!, hash: {:?}", hash_b64_url);
+    stream
+        .for_each_concurrent(10, |msg| {
+            let shared_txn_service = txn_service.clone();
+            let shared_key = keypair.clone();
+            async move {
+                if let Ok(m) = msg {
+                    handle_txn_submission(m, shared_key, shared_txn_service, before_ts).await;
+                } else {
+                    tracing::error!("unable to process msg: {:?}", msg)
+                }
             }
+        })
+        .await;
+    Ok(())
+}
+
+async fn handle_txn_submission(
+    msg: prost::bytes::BytesMut,
+    shared_key: Arc<Keypair>,
+    shared_txn_service: Arc<Mutex<TransactionService>>,
+    before_ts: i64,
+) {
+    if let Ok(Some((txn, hash, hash_b64_url))) = handle_report_msg(msg, shared_key, before_ts) {
+        let mut txn_service = shared_txn_service.lock().await;
```

Comment:
One of the questions I have is why we need a lock on the txn service to serialize txns? The concurrency now means they could already be unordered. 

1. Is the client concurrency safe? if so you don't need a lock
2. If it's not safe you could just create a client per txn? That way the concurrency isn't serialized through a single lock

Reply:
Hmm, fair enough. I mostly just did it this way to get around the rust compiler tbh. So if I understand this correctly, we just clone away txn service for each txn submission? Or create new one via `TransactionService::from_env` and fire it?

Reply:
I believe you can just clone tonic clients since they hold very little state